src/main/java/org/jabref/logic/importer/fetcher/MathSciNet.java

```diff
@@ -102,34 +115,123 @@ public URL getUrlForIdentifier(String identifier) throws URISyntaxException, Mal
+                        for (int i = 0; i < entriesArray.length(); i++) {
+                            JSONObject entryObject = entriesArray.getJSONObject(i);
+                            BibEntry bibEntry = jsonItemToBibEntry(entryObject);
+                            entries.add(bibEntry);
+                        }
+                    } else if (element.has("results")) {
+                        JSONArray entriesArray = element.getJSONArray("results");
+                        for (int i = 0; i < entriesArray.length(); i++) {
+                            JSONObject entryObject = entriesArray.getJSONObject(i);
+                            BibEntry bibEntry = jsonItemToBibEntry(entryObject);
+                            entries.add(bibEntry);
+                        }
                     }
                 }
-            } catch (JSONException | TokenMgrException e) {
+            } catch (JSONException | ParseException e) {
                 LOGGER.error("An error occurred while parsing fetched data", e);
                 throw new ParseException("Error when parsing entry", e);
             }
             return entries;
         };
     }
 
+    private BibEntry jsonItemToBibEntry(JSONObject item) throws ParseException {
+        try {
+            BibEntry entry = new BibEntry(StandardEntryType.Article);
+            // Set fields based on the mappings
+            for (Map.Entry<StandardField, List<String>> mapEntry : fieldMappings.entrySet()) {
+                StandardField field = mapEntry.getKey();
+                List<String> path = mapEntry.getValue();
+
+                String value;
+                if (field == StandardField.AUTHOR) {
+                    value = toAuthors(item.optJSONArray(path.getFirst()));
+                } else if (field == StandardField.KEYWORDS) {
+                    value = getKeywords(item.optJSONObject(path.getFirst()));
+                } else {
+                    value = getOrNull(item, path).orElse(null);
+                }
+
+                if (value != null) {
+                    entry.setField(field, value);
+                }
+            }
+            // Handle articleUrl and mrnumber fields separately
+            String doi = item.optString("articleUrl", "");
+            if (!doi.isEmpty()) {
+                entry.setField(StandardField.DOI, doi);
+            }
+
+            String mrNumber = item.optString("mrnumber", "");
+            if (!mrNumber.isEmpty()) {
+                entry.setField(StandardField.MR_NUMBER, mrNumber);
+            }
+            return entry;
+        } catch (JSONException exception) {
+            throw new ParseException("MathSciNet API JSON format has changed", exception);
+        }
+    }
+
+    private Optional<String> getOrNull(JSONObject item, List<String> keys) {
+        Object value = item;
+        for (String key : keys) {
+            if (value instanceof JSONObject) {
```

Comment:
```suggestion
            if (value instanceof JSONObject obj) {
             ....
```

You can use the new instance of pattern matching syntax,  which makes the extra casting step necessary
https://docs.oracle.com/en/java/javase/17/language/pattern-matching-instanceof-operator.html#GUID-843060B5-240C-4F47-A7B0-95C42E5B08A7

Reply:
Hi, I went through the docs. If I'm not wrong, did you mean that it would make the extra casting step "unnecessary"?

Reply:
I removed the casts and followed the new syntax. It is odd that IntelliJ is giving me red squiggly lines and a suggestion to cast it back. Will ignore for now.

Reply:
Okay nope, I added the obj syntax but seems like the cast is necessary, else I'm getting this compilation error:
error: cannot find symbol
                value = value.opt(key);
                             ^
  symbol:   method opt(String)
  location: variable value of type Object

Reply:
See the examples in the linked docs, you need to use value = obj.get... After the if